### DISCUSSION TITLE
Ending a Kafka Flow programmatically

### DISCUSSION THREAD

**Discussion by shubhanshu02** [bytewax/bytewax#500] [Q&A] [answered]
I am reading the data from a Kafka source where another producer is constantly pushing data. After sometime when all the producing completes, the producer sends a message "exit" to the Kafka and closes the connection. But my flow continuously waits for new messages in the Kafka endlessly.

I want to close the flow when such message is encountered. After this, no processing is required from my ByteWax flow. How do I achieve this behavior?

Right now, I am using the following map operator to quit the program, but this will raise a `SystemExit` exception. What would be the right way to handle this?

```py3
def exit_handler(msg: KafkaMessage):
    if msg['message'] == 'quit':
        exit(0)
    return msg

stream = op.map('exit_handler', stream, exit_handler)
```

**Comment by davidselassie** [comment, marked as answer]
The correct way to handle this is to have a custom input source which monitors for the sentinel "exit" value and raises `StopIteration` to signal EOF to the rest of the dataflow that all data should be flushed and shutdown. One way of describing the dataflow programming model is independent operators communicating only through queues, so there is not way to coherently shut down the dataflow outside of in-band messages and still retain recovery and delivery guarantees; the queues between operators must be flushed for correct answers.

You can copy and modify the existing `KafkaSource` to include this behavior. https://docs.bytewax.io/stable/api/bytewax/bytewax.inputs.html#bytewax.inputs.StatefulSourcePartition.next_batch is the method where `raise StopIteration()` would be added.

I assume this is a single partition topic, but if it's not, because Kafka does not guarantee any delivery ordering across partitions, this system would need to be re-architeced so you can guarantee that you have received all messages before the "exit".
